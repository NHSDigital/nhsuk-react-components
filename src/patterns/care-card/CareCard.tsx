import React, { HTMLProps, createContext, useContext } from 'react';
import classNames from 'classnames';
import { CareCardType } from '../../util/types/NHSUKTypes';
import HeadingLevel, { HeadingLevelType } from '../../util/HeadingLevel';

interface CareCardProps extends HTMLProps<HTMLDivElement> {
  type: CareCardType;
}

const CareCardContext = createContext<CareCardType>('non-urgent');

const genHiddenText = (cardType: CareCardType): string => {
  switch (cardType) {
    case 'non-urgent':
      return 'Non-urgent advice: ';
    case 'urgent':
      return 'Urgent advice: ';
    case 'immediate':
      return 'Immediate action required: ';
    default:
      return '';
  }
};

const CareCardContent: React.FC<HTMLProps<HTMLDivElement>> = ({ className, ...rest }) => (
  <div className={classNames('nhsuk-care-card__content', className)} {...rest} />
);

interface CareCardHeadingProps extends HTMLProps<HTMLHeadingElement> {
  visuallyHiddenText?: false | string;
  headingLevel?: HeadingLevelType;
}

const CareCardHeading: React.FC<CareCardHeadingProps> = ({
  className,
  children,
  visuallyHiddenText,
  headingLevel,
  role,
  ...rest
}) => {
  const cardType = useContext(CareCardContext);
  return (
    <div className="nhsuk-care-card__heading-container">
      <HeadingLevel
        className={classNames('nhsuk-care-card__heading', className)}
        headingLevel={headingLevel}
        {...rest}
      >
        <span role={role}>
          {visuallyHiddenText !== false ? (
            <span className="nhsuk-u-visually-hidden">
              {visuallyHiddenText || genHiddenText(cardType)}
            </span>
          ) : null}
          {children}
        </span>
      </HeadingLevel>
      <span className="nhsuk-care-card__arrow" aria-hidden="true" />
    </div>
  );
};

CareCardHeading.defaultProps = {
  role: 'text',
};

interface CareCard extends React.FC<CareCardProps> {
  Content: React.FC<HTMLProps<HTMLDivElement>>;
  Heading: React.FC<CareCardHeadingProps>;
}

const CareCard: CareCard = ({
  className, type, children, ...rest
}) => (
  <div className={classNames('nhsuk-care-card', `nhsuk-care-card--${type}`, className)} {...rest}>
    <CareCardContext.Provider value={type}>{children}</CareCardContext.Provider>
  </div>
);

CareCard.Content = CareCardContent;
CareCard.Heading = CareCardHeading;

export default CareCard;
