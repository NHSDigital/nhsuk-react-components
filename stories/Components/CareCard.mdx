import {Meta} from "@storybook/blocks";

<Meta title="Components/CareCard" />

# CareCard

This component can be found in the `nhsuk-frontend` repository <a href="https://github.com/nhsuk/nhsuk-frontend/tree/master/packages/components/care-card" target="_blank" rel="noopener noreferrer">here</a>.

## Implementation Notes

By default, CareCard components prepend hidden text before the title. These are:

- ("non-urgent") Non-urgent advice:
- ("urgent") Urgent advice:
- ("immediate") Immediate action required:

If you wish to disable this behaviour, pass the prop `visuallyHiddenText={false}` to the `CareCard.Heading` component or specify your own visually hidden text by using `visuallyHiddenText="Custom"`.

You can change the heading type (i.e. `h1`, `h2` and so on) of the title by passing the prop `headingLevel="<headingLevel>"` to the `CareCard.Heading`.

## Usage


### Standard

```jsx
import { CareCard } from "nhsuk-react-components";

const Element = () => {
    return (
        <CareCard type="non-urgent">
            <CareCard.Heading>Speak to a GP if:</CareCard.Heading>
            <CareCard.Content>
                <ul>
                    <li>you're not sure it's chickenpox</li>
                    <li>the skin around the blisters is red, hot or painful (signs of infection)</li>
                    <li>your child is <a href="https://www.nhs.uk/conditions/dehydration">dehydrated</a></li>
                    <li>you're concerned about your child or they get worse</li>
                </ul>
                <p>Tell the receptionist you think it's chickenpox before going in. They may recommend a special appointment time if other patients are at risk.</p>
            </CareCard.Content>
        </CareCard>
    );
}
```
