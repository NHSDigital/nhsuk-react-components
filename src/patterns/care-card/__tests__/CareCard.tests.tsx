import React from 'react';
import { mount, shallow } from 'enzyme';
import CareCard from '..';

describe('CareCard', () => {
  it('matches snapshot', () => {
    const component = mount(<CareCard type="urgent" />);
    expect(component).toMatchSnapshot('BaseCareCard');
    component.unmount();
  });

  it('generates the correct hidden text', () => {
    const nonUrgentCard = mount(
      <CareCard type="non-urgent">
        <CareCard.Heading>Heading</CareCard.Heading>
      </CareCard>,
    );
    const urgentCard = mount(
      <CareCard type="urgent">
        <CareCard.Heading>Heading</CareCard.Heading>
      </CareCard>,
    );
    const immediateCard = mount(
      <CareCard type="immediate">
        <CareCard.Heading>Heading</CareCard.Heading>
      </CareCard>,
    );

    const withoutHiddenText = mount(
      <CareCard type="immediate">
        <CareCard.Heading visuallyHiddenText={false}>Heading</CareCard.Heading>
      </CareCard>,
    );

    const withCustomHiddentext = mount(
      <CareCard type="immediate">
        <CareCard.Heading visuallyHiddenText="Custom">Heading</CareCard.Heading>
      </CareCard>,
    );

    expect(nonUrgentCard.find('.nhsuk-u-visually-hidden').text()).toEqual('Non-urgent advice: ');
    expect(urgentCard.find('.nhsuk-u-visually-hidden').text()).toEqual('Urgent advice: ');
    expect(immediateCard.find('.nhsuk-u-visually-hidden').text()).toEqual(
      'Immediate action required: ',
    );
    expect(withCustomHiddentext.find('.nhsuk-u-visually-hidden').text()).toEqual('Custom');
    expect(withoutHiddenText.find('.nhsuk-u-visually-hidden').exists()).toBeFalsy();

    nonUrgentCard.unmount();
    urgentCard.unmount();
    immediateCard.unmount();
    withoutHiddenText.unmount();
    withCustomHiddentext.unmount();
  });

  it('renders with correct classNames', () => {
    const nonUrgentCard = shallow(<CareCard type="non-urgent" />);
    const urgentCard = shallow(<CareCard type="urgent" />);
    const immediateCard = shallow(<CareCard type="immediate" />);

    expect(nonUrgentCard.hasClass('nhsuk-care-card--non-urgent')).toBeTruthy();
    expect(urgentCard.hasClass('nhsuk-care-card--urgent')).toBeTruthy();
    expect(immediateCard.hasClass('nhsuk-care-card--immediate')).toBeTruthy();

    nonUrgentCard.unmount();
    urgentCard.unmount();
    immediateCard.unmount();
  });

  it('renders content', () => {
    const component = mount(
      <CareCard type="non-urgent">
        <CareCard.Content>Test Content</CareCard.Content>
      </CareCard>,
    );
    expect(component.find('.nhsuk-care-card__content').text()).toEqual('Test Content');
    component.unmount();
  });
});
