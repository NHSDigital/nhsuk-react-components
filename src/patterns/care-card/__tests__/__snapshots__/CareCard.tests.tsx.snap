// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CareCard matches snapshot: BaseCareCard 1`] = `
<CareCard
  type="urgent"
>
  <div
    className="nhsuk-care-card nhsuk-care-card--urgent"
  />
</CareCard>
`;
